### PR TITLE
Added summarizer on server side

### PR DESCRIPTION
**Pull request by Georgehe4**
Summarization moved over to services.

### CODE REVIEW

**Comment by Georgehe4** [comment]
#226 
@devoncarew @lukechurch

**Comment by lukechurch** [comment]
Lets talk on Monday about the various choices of how to expose this in the APIs.

File: lib/src/common_server.dart [hunk 1]
```diff
@@ -248,6 +258,20 @@ class CommonServer {
     }
     return _analyzeMulti({"main.dart" : source});
   }
+  
+  Future<AnalysisResults> _summarize(String dart, String html, String css) async {
+    if (dart == null || html == null || css == null) {
```
Comment (lukechurch) [inline lib/src/common_server.dart]: We've tended to call the thing that is being called 'dart' here, 'source' elsewhere in the server.

Could we go for either source or dartSource?

File: lib/src/summarize.dart [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+class Summarizer {
+  final String dart;
+  final String html;
+  final String css;
+  final AnalysisResults analysis;
+
+  _SummarizeToken storage;
+  int _randomizer;
+
+  static Map<String, List<int>> cuttoffs = {
```
Comment (lukechurch) [inline lib/src/summarize.dart]: Any reason for these not to be final?

File: lib/src/summarize.dart [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+      'very few instances of'
+    ],
+    'errorCount-0': ['zero', 'no', 'a nonexistent amount of', '0'],
+    'use': ['demonstrates', 'illustrates', 'depicts'],
+    'code-0': ['it'],
+    'code-1': ['It'],
+  };
+
+  Summarizer({this.dart, this.html, this.css, this.analysis}) {
+    if (dart == null) throw new ArgumentError('Input cannot be null.');
```
Comment (lukechurch) [inline lib/src/summarize.dart]: Why is dart it an optional parameter?

File: lib/src/summarize.dart [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+  List<String> _codeSearch() {
+    List<String> features = new List<String>();
+    for (String feature in codeKeyWords.keys) {
+      if (_usedInDartSource(feature)) features.add(codeKeyWords[feature]);
+    }
+    return features;
+  }
+
+  String _featureList(List<String> list) {
+    if (list.length == 0) return '. ';
```
Comment (lukechurch) [inline lib/src/summarize.dart]: Stringbuffers!

**Comment by lukechurch** [comment]
Thanks @Georgehe4 I know you're busy with embedding stuff, if you need me to pick up the code changes here, let me know.

File: lib/src/common_server.dart [hunk 5]
```diff
@@ -248,6 +258,20 @@ class CommonServer {
     }
     return _analyzeMulti({"main.dart" : source});
   }
+  
+  Future<AnalysisResults> _summarize(String dart, String html, String css) async {
```
Comment (devoncarew) [inline lib/src/common_server.dart]: If you don't use `await` in the method body then you don't need `async` on the method declaration.

**Comment by devoncarew** [comment]
@lukechurch, just an fyi related to some of your review comments, I believe `summarize.dart` and `summarize_test.dart` were moved over wholesale from `dart-pad`. Only the changes to `common_server.dart` are new code.

**Comment by devoncarew** [comment]
Also, if you reference an issue number, it'll create a link to that issue. If you reference is like (fix #my_num), it'll automatically close the issue when the PR is merged. Pretty handy. https://help.github.com/articles/closing-issues-via-commit-messages/ for the full list of shortcuts available.

**Comment by lukechurch** [comment]
@devoncarew I see. Well, as soon as we've agreed on the end point, and run the extended test suite we can bring this in and iterate.

**Comment by lukechurch** [comment]
I'm going to bring this in and fix it up before the 1.11 push.

**Pull request merged**